### ISSUE TITLE
[Migration] Torchaudio Complex Tensor Support and Migration

### ISSUE THREAD

**Issue by mthrok** [pytorch/audio#1337]
# Torchaudio Complex Tensor Support and Migration

## Overview

`torchaudio` has been expressing complex numbers by having an extra dimension for real-part and imaginary-part. (We will refer this format as `"pseudo complex type"`)

```python
waveform = torch.randn(...)
spectrogram = torchaudio.functional.spectrogram(waveform, ..., power=None)

# the last dimension represents real and imaginary parts of complex tensors.
print(spectrogram.dtype, spectrogram.shape)
>>> torch.float32, torch.Size([... ,2])
```

PyTorch 1.6 introduced complex Tensor type, such as `torch.complex64` (`torch.cfloat`) and `torch.complex128` (`torch.cdouble`). (Will be refered as `"native complex type"`)

The natitve complex type comes with handy methods for complex operation such as `abs`, `angle` and `magphase`. (Please refer to the official documentation for the detail.)

Over the few coming releases, we plan to migrate torchaudio's functions and transforms to the native complex type. This issues describes the planned approaches/works/changes/timeline. If you have a question, a concern or a suggestion. Feel free to leave a comment.

## Migration Stages

We will perform the migration in multiple stages. At this moment, the completion of the later migration stages re not tied with specific releases yet.

### ✅ Stage 0 (`~ 0.8`)
Up to release 0.8, torchaudio exclusively used pseudo complex type. In PyTorch 1.7, PyTorch started the adaptation of native complex type and the migration of `torch.fft` namespace. Because of this, torchaudio already uses native complex type in some implementations (`F.vad`, `T.Vad`, `kaldi.spectrogram` and `kaldi.fbank`) but all the user facing APIs use use pseudo complex type.

### ✅ Stage 1 Add support for native complex type and deprecate pseudo complex type)

**Completed:** PyTorch 1.9 / torchaudio 0.9

#### Library code change
In this stage, torchaudio will support both pseudo complex type and native complex type. This means that

* Functions that accept complex input should be able to handle both pseudo and native complex types.
* Functions that return complex values can return both pseudo and native complex types.
    * For real-to-complex functions, a new argument `return_complex` will be added so that users can switch the behavior.

#### Test code update

In addition to the above library code changes, we are going to add a set of tests to make sure that native complex types work in common use cases. This includes;

* Gradient check
* JIT
* Performance 
* Distributed training
* `nn.Module` compatibility

### ✅Stage 2 (Switch to native complex type by default)

**Completed:** `main` branch. To be released as part of PyTorch 1.10 / torchaudio 0.10

The default value for `return_complex` is changed to `True`.

### 👉 Stage 3 (Remove the support for pseudo complex type)

In this stage, we will remove the support for pseudo complex type.

* Functions that work with complex types should handle native complex types exclusively.
* Passing pseudo complex type results in an error.
* `return_complex` argument added in Stage. 1 is deprecated and eventually removed.

## Affected Functions
The following figure illustrates the functions that handle complex values and their dependencies.

<img width="1365" alt="Screen Shot 2021-03-02 at 10 59 30" src="https://user-images.githubusercontent.com/855818/109676241-8426e200-7b46-11eb-884e-6fa5dc5c5b43.png">

#### `F.angle`,  `F.complex_norm`, `T.ComplexNorm`, `F.magphase `
These functions are deprecated in Stage.1 and will be removed in Stage.3.
For `F.angle`, native complex tensors provide the `angle()` function.
For `F.complex_norm` / `T.ComplexNorm`, the equivalent computation can be performed with `abs().pow(n)`.
`F.magphase` is a convenient function to call `F.angle` and `F.complex_norm`, therefore, this function is deprecated as well.

#### `F.griffinlim`, `T.GriffinLim`
Changes to these functions are kept internal, therefore we can simply change the internals without disturbing the downstream users.

#### `F.phase_vocoder`, `T.TimeStretch`
When adding support for native complex type, we can simplify the interface change as follow

* If the input is pseudo complex type, return pseudo complex type
* If the input is native complex type, return native complex type

#### `F.spectrogram`, `T.Spectrogram`
These functions return either real valued Tensor (power, energy) or complex valued Tensor (frequency representation), which depends on what `power` argument was provided. When `power` is not provided, these functions return a complex-valued Tensor. In this case, users have the option to receive the result in pseudo complex type or native complex type. `return_complex` argument will be added for this choice. If `return_complex` is `True`, then native complex type is returned. See #1009 for the discussion.

## Timeline

<table>
 <tr>
  <td></td>
  <td><b>Migration Phase</b></td>
  <td align=center>1</td>
  <td align=center>2</td>
  <td align=center>3</td>
 </tr>
 <tr>
  <td></td>
  <td><b>PyTorch</b>/<b>torchaudio</b> versions</td>
  <td align=center>1.9 / 0.9</td>
  <td align=center>1.10 / 0.10</td>
  <td align=center>TBD</td>
 </tr>
 <tr>
  <td><b>Class / function</b></td>
  <td><b>Type</b></td>
  <td colspan=3</td>
 </tr>
 <tr>
  <td><code>F.angle</code>,
  <code>F.complex_norm</code>,
  <code>F.magphase</code>,
  <code>T.ComplexNorm</code></td>
  <td>C-&gt;R, utility</td>
  <td>Deprecated</td>
  <td>Deprecated</td>
  <td>Removed</td>
 </tr>
 <tr>
  <td><code>F.griffinlim</code>, <code>T.GriffinLim</code></td>
  <td>R-&gt;R</td>
  <td>Adopts native complex internally</td>
  <td colspan=2>No change</td>
 </tr>
 <tr>
  <td><code>F.phase_vocoder</code>, <code>T.TimeStretch</code></td>
  <td>C-&gt;C</td>
  <td>
    <ul>
      <li>Support for native complex type is added</li>
      <li>Support for pseudo complex type is deprecated.</li>
      <li>The function returns the same type as the input.<br>(native for native, pseudo for pseudo)</li>
    </ul></td>
  <td></td>
  <td>Support for pseudo complex type is removed.<br/>Only handles native complex type.</td>
 </tr>
 <tr>
  <td><code>F.spectrogram</code>, <code>T.Spectrogram</code></td>
  <td>R-&gt;C (when <code>power=None</code>)</td>
  <td>Argument <code>return_complex</code> is added. (default value is <code>False</code>)<br>
    When the return value is complex-valued (<code>power=None</code>),<br>
    the type of the returned Tensor can be switched with <code>return_complex</code>.</td>
  <td>The default value of <code>return_complex</code> is changed to <code>True</code>.</td>
  <td>The <code>return_complex</code> argument is deprecated.</td>
 </tr>
</table>

### `F.angle`, `F.complex_norm`, `F.magphase` and `T.ComplexNorm`

<table>
<tr>
  <td>~0.8</td>
  <td>0.9~</td>
</tr>
<tr>
  <td>

```python

spectrogram = ...  # Tensor with pseudo complex type (shape == (..., 2))
angle = F.angle(spectrogram)
magnitude = F.complex_norm(spectrogram, norm=1)
power = F.complex_norm(spectrogram, norm=2)
norm = F.complex_norm(spectrogram, norm=norm)
magnitude, phase = F.magphase(spectrogram, n)
```

  </td>
  <td>

```python
spectrogram = ...  # Tensor with pseudo complex type (shape == (..., 2))
spectrogram = torch.view_as_complex(spectrogram)  
angle = spectrogram.angle()
magnitude = spectrogram.abs()
power = spectrogram.abs().pow(2)
norm = spectrogram.abs().pow(norm)
magnitude, phase = spectrogram.abs().pow(n), spectrogram.angle()
```

  </td>
</tr>
</table>

### `F.phase_vocoder`, `T.TimeStretch`

<table>
<tr>
  <td>~0.8</td>
  <td>0.9~</td>
</tr>
<tr>
  <td>

```python
spec = ... # pseudo complex (..., 2) 

## If using functional form
spec = F.phase_vocoder(spec, ...)
## else using transform
transform = T.TimeStretch(...)
spec = transform(spec)
```

  </td>
  <td>

```python
# Convert to native complex
spec = ... # pseudo complex (..., 2) 
# convert to native complex type
spec = torch.view_as_complex(spec)

# Perform the operation
## If using functional form
spec = F.phase_vocoder(spec, ...)
## else using transform
transform = T.TimeStretch(...)
spec = transform(spec)

# Convert back to pseudo complex type
# (If your downstream code still expects pseudo complex type)
spec = torch.view_as_real(spec)
```

  </td>
</tr>
</table>

### `F.spectrogram`, `T.Spectrogram`

<table>
<tr>
  <td>~0.8</td>
  <td>0.9~</td>
</tr>
<tr>
  <td>

```python
spec = F.spectrogram(waveform, ..., power=None)

transform = T.Spectrogram(..., power=None)
spec = transform(waveform)  # pseudo complex (..., 2) 
```

  </td>
  <td>

```python
spec = F.spectrogram(waveform, ..., power=None, return_complex=True)

transform = T.Spectrogram(..., power=None, return_complex=True)
spec = transform(waveform)  # native complex
# If your downstream code still expects pseudo complex type
spec = torch.view_as_real(spec)
```

  </td>
</tr>
</table>

##### Code Change
- [x] `F.griffinlim`
    #1368
- [x] `F.phase_vocoder`, `T.TimeStretch`
    #1410
    ~#758~ 
- [x] `F.spectrogram`, `T.Spectrogram`, ~`T.MelSpectrogram`~
    #1366 
    ~#1009~ 
##### Add deprecation Warnings
- [x] `F.angle`, `F.complex_norm`, `T.ComplexNorm`, `F.phase_vocoder`, `T.TimeStretch`, `F.spectrogram`, `T.Spectrogram`
    #1445
    ~#1431~
- [x] `F.magphase` #1492
##### Change the default value of `return_complex` to `True`.
- [x] `F.spectrogram`, `T.Spectrogram` #1549
##### Update the deprecation warnings to indicate the version of removal.
- [x] `F.angle`, `F.complex_norm`, `F.magphase`, `F.phase_vocoder`, `T.TimeStretch`, `T.ComplexNorm` #1553
##### Remove the support for pseudo complex type.
- [x] Remove `F.magphase`  #1934 
- [x] Remove `F.angle` #1935 
- [x] Remove `F.complex_norm` and `T.ComplexNorm` #1942 
- [x] Remove support for pseudo complex type from `F.phase_vocoder` and `T.TimeStretch` #1957
- [x] Remove support for pseudo complex type from `F.spectrogram` and `T.Spectrogram`  #1958 
    - [x] Deprecate `return_complex` argument from `F.spectrogram` and `T.Spectrogram`
#### Conjugate input tests
- [ ] After https://github.com/pytorch/pytorch/pull/54987, check if things work when the input complex Tensor is conjugate.

#### Autograd tests
- [x] #1421 `T.GriffinLim`
    - [ ] [Optional] `F.griffinlim`
-  [x] #1420 `T.TimeStretch`, `F.phase_vocoder`
    **NOTE** They are not differentiable around zero.
- [x] #1340 `T.Spectrogram`, `T.MelSpectrogram`
    - [x] Investigate why `T.Spectrogram` is not deterministic on its backward pass.  [[error](https://app.circleci.com/pipelines/github/pytorch/audio/5301/workflows/782e2d49-6f99-4e98-a240-b6597557d855/jobs/176159) -> [issue pytorch/#54093](https://github.com/pytorch/pytorch/issues/54093)]
    - [ ] [Optional] `F.spectrogram`

#### Ensuring TorchScript support
Check if all the functionals/transforms are covered by TorchScript consistency test and add if missing
* [`F.spectrogram`](https://github.com/pytorch/audio/blob/ea857940de9e3738166989ad3bf1726741a13f04/test/torchaudio_unittest/functional/torchscript_consistency_impl.py#L23-L35)
* [`F.griffinlim`](https://github.com/pytorch/audio/blob/ea857940de9e3738166989ad3bf1726741a13f04/test/torchaudio_unittest/functional/torchscript_consistency_impl.py#L37-L51)
* [`F.phase_vocoder`](https://github.com/pytorch/audio/blob/0b1d1f06595ec7897270f6b924aa76dcc13ffd7b/test/torchaudio_unittest/functional/torchscript_consistency_impl.py#L550-L563)
* [`T.Spectrogram`](https://github.com/pytorch/audio/blob/ea857940de9e3738166989ad3bf1726741a13f04/test/torchaudio_unittest/transforms/torchscript_consistency_impl.py#L20-L222)
* [`T.GriffinLim`](https://github.com/pytorch/audio/blob/ea857940de9e3738166989ad3bf1726741a13f04/test/torchaudio_unittest/transforms/torchscript_consistency_impl.py#L24-L26)
* [`T.MelSpectrogram`](https://github.com/pytorch/audio/blob/ea857940de9e3738166989ad3bf1726741a13f04/test/torchaudio_unittest/transforms/torchscript_consistency_impl.py#L36-L38)
* [`T.TimeStretch`](https://github.com/pytorch/audio/blob/ea857940de9e3738166989ad3bf1726741a13f04/test/torchaudio_unittest/transforms/torchscript_consistency_impl.py#L61-L69)

- [x] #1446 Add step to dump TorchScripted object in TorchScript test.

#### Benchmark

- `F.phase_vocoder` #1410 x6 speed up on CPU. Not much different on GPU

cc @anjali411 @vincentqb

**Comment by JuanFMontesinos** [comment]
Please, before going forward with the deprecation, note that complex32 format is poorly supported on cuda.
Cannot do complex32 product, stft and istft allows power2 kernels only and a large list of drawbacks. 

Had to rewrite whole code back to pseudo complex due to this.

**Comment by mthrok** [comment]
> Cannot do complex32 product, stft and istft allows power2 kernels only and a large list of drawbacks.

Hi @JuanFMontesinos 

Thanks for letting us know. 

TorchAudio is not tested on fp16 nor complex32, and they are not part of officially supported types. So we did not realize that pseudo complex can be used for a workaround of complex32.

Unfortunately, we wrapped up the release v0.11 (scheduled to be out in about one week) and PyTorch removed complex32 type and torchaudio removed the support for pseudo complex type. So I assume it will be unusable to you. I can try reverting the pseudo complex support if that's the best course of action. (however it is known that some operations with pseudo complex have issues with accuracy, so it's not the best workaround, which is one we wanted to migrate to native complex type.)

The treatment of complex32 indeed needs improvement and there is an issue created for this in PyTorch. https://github.com/pytorch/pytorch/issues/71680 The most ideal outcome is that PyTorch core adds complex32 support quickly but I am not sure if that can happen quickly. However, your voice matters a lot here, so would you be willing to provide features that will be most relevant for you in https://github.com/pytorch/pytorch/issues/71680? That way, the PyTorch core team can prioritize it if they decide to work on them.
